Camera controls extension

DT-is-not-available opened the pull request:
Allows free movement of the camera along the X and Y axis, and zooming. Note that this extension automatically removes fencing to make this work correctly.
Adds the following blocks:
![image](https://user-images.githubusercontent.com/66632141/221392337-6e3d9fb5-7a64-419b-8d93-b636c6aa374d.png)
I currently do not have an image for the extension, I will make one sometime tomorrow.

TheShovel commented:
This is something I wanted to see for a while now. This is cool.

True-Fantom commented:
@DT-is-not-available

Hello! I have slightly changed your extension so that it is in scratch style:

![27-02-2023 160146](https://user-images.githubusercontent.com/116278271/221571338-8df7eff0-61e4-43b9-beee-8fa4528c3da1.png)

And also I added icon and created new image (#204)
![81](https://user-images.githubusercontent.com/116278271/221571868-337fa663-c379-42a5-9c56-cd62f27367c9.svg)
![82](https://user-images.githubusercontent.com/116278271/221571871-d2a2af9d-e0b7-4dd2-b36f-a360bc166a34.svg)
[Here](https://github.com/True-Fantom/true-fantom.github.io/blob/main/tests/camera.js) is the finished code, just copy it

DNin01 commented:
Would this coincide with Box2D?

True-Fantom commented:
@DNin01 I looked at the code, apparently yes, but it's not accurate

DT-is-not-available commented:
@True-Fantom 
Thanks! I will update the images and credit you later today. As for the palette layout, I had no idea you could just insert "---" as a separator, very useful.

@DNin01
I have no idea, I'll have to check but I imagine it wouldn't work perfectly if at all.

True-Fantom commented:
@DT-is-not-available Was happy to help)

DT-is-not-available commented:
I've gone ahead and added the images to the extension, as well as the modified block palette. I do have an unused "set background color" block, since it doesn't change the result of the touching-color boolean. Does anyone know how to do that? Also at some point I want to add camera rotation, but I'm not sure how to do that at this point, I'll see if I can tell webgl to apply proper transformations before drawing anything rather than messing with the projection values.

GarboMuffin commented:
It helps to post the code that doesn't work when asking for further help

GarboMuffin commented:
Ok it is in there, just a bit indirect

Look at vm.renderer.setBackgroundColor

DT-is-not-available commented:
Oh I see now, should look through scratch-render some more, thanks!

True-Fantom commented:
@DT-is-not-available Since you added a background color, I think need to add reporter for it:
![154313](https://user-images.githubusercontent.com/116278271/222432986-43bfd6f2-be03-4538-8fd9-8ea23970429f.png)
+position the block a little higher) [Here](https://github.com/True-Fantom/true-fantom.github.io/blob/main/tests/camera.js) is code

DT-is-not-available commented:
Background color feels unrelated to camera movement, so I'll probably move it and a reporter under the movement blocks. (I may add more visual blocks since proper transformations requires custom shaders, so I might as well use them)

True-Fantom commented:
Ok, that's a great idea

DT-is-not-available commented:
I will modify the layout of the extension tonight, and probably leave it as is for a while until I learn more about how webgl works. On that note, is there a way I should go about applying custom glsl code as a vertex shader without screwing up the default?

GarboMuffin commented:
Why does this extension have unused code related to shaders?

ghbbeep commented:
I'm assuming they're not talking about this extension.

GarboMuffin commented:
The extension in this PR appears to contain a lot of code and shaders that are seemingly unused. Is that intentional?

DT-is-not-available commented:
@GarboMuffin That was not intentional, my bad. I am considering rewriting the extension to use a custom shader however, as this would allow for much easier implementation of things like camera rotation, and well, camera "shader" effects. I've removed the shader related code and updated the set background color block to be at the end of the block palette, along with a background color reporter.

DT-is-not-available commented:
Alright, I'm done updating this for a while.

True-Fantom commented:
@GarboMuffin Can you merge this, please?
